apps/passport-server/test/generic-issuance/credentialSubservice.spec.ts

```diff
@@ -0,0 +1,177 @@
+        await credentialSubservice.verifyAndExpectZupassEmail(credential);
+      expect(verifiedCredentialWithEmail).to.deep.eq(verifiedCredential);
+    }
+    {
+      const credential = await makeCredential(
+        identity,
+        // Credentials for Zupass feeds do not include Email PCDs
+        ZUPASS_CREDENTIAL_REQUEST
+      );
+
+      // Same promise will be returned for the same input
+      const verifyPromise = credentialSubservice.verify(credential);
+      const secondPromise = credentialSubservice.verify(credential);
+      expect(verifyPromise).to.eq(secondPromise);
+
+      // Result is a verified credential containing the expected values
+      const verifiedCredential = await verifyPromise;
+      expect(verifiedCredential.signatureClaim.identityCommitment).to.eq(
+        identity.getCommitment().toString()
+      );
+      expect(verifiedCredential.emailClaim).to.be.undefined;
+      expect(verifiedCredential.emailSignatureClaim).to.be.undefined;
+
+      // Verifying this with the expectation of a valid email should throw,
+      // since the credential does not have an EmailPCD.
+      try {
+        expect(
+          await credentialSubservice.verifyAndExpectZupassEmail(credential)
+        ).to.throw;
+      } catch (e) {
+        expect(e instanceof VerificationError).to.be.true;
+      }
+    }
+  });
+
+  /**
+   * Very similar tests are carried out in Lemonade/Pretix pipelines, but these
+   * tests focus specifically on the fact that CredentialSubservice should
+   * throw an exception.
+   */
+  step("credential subservice won't verify invalid credentials", async () => {
+    const emailAddress = "test@example.com";
+    const identity = new Identity();
+    const badEmailPCD = await proveEmailPCD(
+      emailAddress,
+      // Not the Zupass private key!
+      newEdDSAPrivateKey(),
+      identity
+    );
+    const goodEmailPCD = await proveEmailPCD(
+      emailAddress,
+      ZUPASS_EDDSA_PRIVATE_KEY,
+      identity
+    );
+    // Credential containing an invalid EmailPCD
+    const badEmailCredential = await signCredentialPayload(
+      identity,
+      createCredentialPayload(await EmailPCDPackage.serialize(badEmailPCD))
+    );
+    const mismatchedIdentityCredential = await signCredentialPayload(
+      // Semaphore identity is different from that used by the Email PCD
+      new Identity(),
+      createCredentialPayload(await EmailPCDPackage.serialize(goodEmailPCD))
+    );
+    MockDate.set(Date.now() - ONE_DAY_MS);
+    const expiredCredential = await signCredentialPayload(
+      identity,
+      createCredentialPayload(await EmailPCDPackage.serialize(goodEmailPCD))
+    );
+    MockDate.set(Date.now() + ONE_DAY_MS);
+
+    const badCredentials = [
+      badEmailCredential,
+      mismatchedIdentityCredential,
+      expiredCredential
+    ];
+
+    for (const credential of badCredentials) {
+      try {
+        expect(await credentialSubservice.verify(credential)).to.throw;
```

Comment:
im not sure this is doing what you think it's doing - usually assertions about promises themselves need to be awaited, or the assertion should be something like `to.eventually.throw` (i forget the precise syntax). what's happening here I think is that the `expect(...).to.throw` is effectively a no-op, because `credentialSubservice.verify` throws which gets immediately caught on line 172.

Reply:
Couldn't get `eventually` to work, but I've replaced it with a different means of testing the same thing.